Creating "FusionDesk (React SDK)" blog post (#2029)
FusionAuth/fusionauth-site: site/_posts/2023-03-09-react-example-application.md

---
layout: blog-post
title: Using the FusionAuth React SDK in a React application
description: In this guide, we'll showcase how to use the FusionAuth React SDK to implement the login, logout, and registration flows in a React application.
author: Colin Frick
category: blog
image: blogs/react-example-application/react-example-application.png
tags: tutorial tutorial-react tutorial-typescript react typescript oauth javascript
excerpt_separator: "<!--more-->"
---

The React SDK from FusionAuth enables developers to implement the login, logout, and registration flows in a React application with ease.

<!--more-->

> To show the React SDK in action, we will be using the FusionDesk application. FusionDesk is a simple help desk example application that allows users to create tickets and view them. The application is built using React and Node.js. It uses a server backend built using Ts.Ed, Node.js, and Typescript. The FusionDesk application is available on GitHub.

Reviewer: "available on GitHub" should be linked to the GH repo. That should be transferred to the FusionAuth org (or just give me read rights and I can move it).

Reviewer: Ah, I see you added it below. I'd remove "The FusionDesk application is available on GitHub." because you show that in a few paragraphs.

{% include _image.liquid src="/assets/img/blogs/react-example-application/fusiondesk-tickets.png" alt="FusionDesk showing the main list of tickets" class="img-fluid" figure=false %}

### Prerequisites

To follow along with this article, you will need to have the following installed:

* Node.js
* Git
* Docker with Docker Compose (optional)

{% include _callout-tip.liquid
content=

> "If you want to run the example application on another domain than `localhost`, you need to make sure that FusionAuth runs on the same domain as the backend of the application. This is a limitation of the Hosted OAuth Service Provider Endpoints."

Reviewer: I don't understand this. Can you provide an example?

Author: Because the Hosted OAuth Service Provider Endpoints set the access_token cookies on the domain where FusionAuth is hosted, the back end must be hosted on the same domain - or sub domain.
F.e on my dev machine I set up the demo application like this:
Frontend: `desk.fusionauth.io`
Backend: `api.fusionauth.io`
FusionAuth: `local.fusionauth.io`

Reviewer: That's great. Please include that. Please use the example.com domain.

%}

{% include _callout-important.liquid
content=

> "This article was written using an early look at the FusionAuth Hosted OAuth Service Provider Endpoints. The feature may be not yet available in the latest version of FusionAuth."

Reviewer: Point to the GH issue that is tracking that so someone can check it out. https://github.com/FusionAuth/fusionauth-issues/issues/1943

%}

### Clone the Repository

To begin, clone the demo application from GitHub using the following command:

```bash

> git clone https://github.com/FusionAuth/fusionauth-react-example-app-placeholder-replaceme

Reviewer: What is the real URL of the repo?

```

### Set up the FusionAuth application

To set up the FusionAuth application, follow the steps below:

```bash
cd fusionauth-react-example-app-placeholder-replaceme
docker compose up -d
# Wait until FusionAuth is set up and available at http://localhost:9011
cd server && cp example.env .env && npm install && npm run seed && cd ..
cd client && cp example.env .env && npm install && cd ..
```

### Run the Application

To run the application, follow the steps below:

```bash
cd server && npm run start
```

Open a new terminal and run the following command:

```bash
cd client && npm run start
```

The application is now running on `http://localhost:3000`.

Use the following credentials to log in:
- Email: `admin@example.com`
- Password: `password`

This will log you in as an agent. You can also log in as a user by using the following credentials:
- Email: `richard@example.com`
- Password: `password`

> ### Set up the Fusionauth Instance Manually

Reviewer: ```suggestion
### Set up the Fusionauth Instance manually
```

For blog posts, we don't capitalize anything except proper nouns.

Reviewer: Please review and update all the headers.

If you do not want to use Docker, you can set up the FusionAuth instance manually. To do so, follow the steps detailed in the [FusionAuth documentation](https://fusionauth.io/docs/v1/tech/installation-guide/fusionauth-app) and the [README.md](https://github.com/FusionAuth/fusionauth-react-example-app-placeholder-replaceme/blob/main/README.md) of the example application.

## Using Hosted OAuth Service Provider Endpoints

Authentication is hard. We want to make it easy.

> That is why we have added the necessary endpoints for the OAuth 2.0 login flow into the FusionAuth API. This means that you can use FusionAuth to authenticate users. This is a great way to get started with FusionAuth and to use it as a drop-in replacement for your existing OAuth 2.0 provider. So you can concentrate on providing the best experience for your users.

Reviewer: ```suggestion
That is why we have added the necessary endpoints for the OAuth 2.0 login flow into the FusionAuth API. This means that you can use FusionAuth to authenticate users without standing up a separate back end to handle the token exchange. This is a great way to get started with FusionAuth and to use it as a drop-in replacement for your existing OAuth 2.0 provider. So you can concentrate on providing the best experience for your users.
```

{% include _callout-important.liquid
content=
"For limitations and prerequisites, see the [Hosted OAuth Service Provider Endpoints](https://fusionauth.io/xyz) documentation."
%}

{% include _image.liquid src="/assets/img/blogs/react-example-application/fusiondesk-fusionauth-login.png" alt="Login Screen of FusionAuth with the styling of FusionDesk" class="img-fluid" figure=false %}

> As you can see - if you inspect the backend of the example application - we do not provide any OAuth 2.0 endpoints. Instead, we use the FusionAuth API to authenticate users.

Reviewer: ```suggestion
As you can see - if you inspect the backend of the example application - we do not provide any OAuth 2.0 endpoints. Instead, we use FusionAuth provided functionality to authenticate users.
```

You aren't really using the FusionAuth API, are you?

> When the user accesses the backend, we validate the `access_token` cookie. If the cookie is valid, we will return the tickets. If the cookie is not valid, we return a 401 error.

Reviewer: Feels like you should talk about how a user logs in before you talk about how the app functionality is accessed.

> See `server/src/middlewares/fusion-auth.middleware.ts` for the implementation.

Reviewer: Let's put some code in here. You can use the `remote_include` plugin.

For example: `{% remote_include 'https://raw.githubusercontent.com/FusionAuth/fusionauth-example-client-libraries/main/java/src/main/java/io/fusionauth/example/AddMagicLink.java' %}`

{% include _callout-note.liquid
content=

> "Currently every request validates the `access_token` with FusionAuth using the `/oauth2/introspect` endpoint. This is not the most efficient way to validate the token. Depending on the security requirements of your application, you might want to cache this information or validate the `access_token` directly without using the FusionAuth API."

Reviewer: Can you change this to use the JWKS endpoint and validate the JWT locally? Here's an example: https://github.com/FusionAuth/fusionauth-example-modern-guide-to-oauth/blob/master/routes/common.js#L21

%}

> We also use the FusionAuth API to retrieve user information for the creator of a ticket. This is done in the `TicketController.ts` file using the `retrieveUser` method of the FusionAuth API client for typescript.

Reviewer: Add some code, please

### Theming the Login / Register View in Fusionauth

> The login / register view in FusionAuth can be themed using the FusionAuth theme editor. You can find more information about the theme editor in the [FusionAuth documentation](https://fusionauth.io/docs/v1/tech/themes/).

Reviewer: ```suggestion
The login / register view in FusionAuth can be themed using the FusionAuth theme editor. You can find more information about the theme editor in the [FusionAuth documentation](/docs/v1/tech/themes/).
```

Please keep links relative.

> For this example application, we integrate Tailwind CSS and DaisyUI into the theme by automatically generating the css stylesheet based on the FusionAuth theme templates.

Reviewer: I want to know more about this. Can you put in a few more paragraphs about how you automatically generate the CSS from the theme templates?

> ## Benefits SDK

Reviewer: ```suggestion
## Benefits of the React SDK
```

The React SDK enables developers to implement the login, logout, and registration flows in a React application.

It provides pre-built buttons for login, logout, and registration. It also provides a `withFusionAuth` [HOC](https://reactjs.org/docs/higher-order-components.html) that can be used to wrap a component and provide the user information to the component.
With `RequireAuth` you can wrap a component and require the user to be authenticated before the component is rendered. The `withRole` option also enables you to restrict the user to a specific role.

> `RequireAuth` is used in the `LoggedInMenu` component, which is used to display the user information, profile, and logout button if the user is authenticated.

Reviewer: Would be a great place to show code.

```tsx
export const ProtectedRoutes: FC = () => {
  const {isLoading, isAuthenticated} = useFusionAuth();

  if (isLoading) return null;

  return (isAuthenticated ? <Outlet/> : <Navigate to={'/'} replace={true}/>);
};
```

{% include _callout-note.liquid
content=
"For more information about the React SDK, see the [FusionAuth React SDK](/docs/v1/tech/client-libraries/react-sdk) documentation."
%}

{% include _callout-note.liquid
content=
"Any code snippets in this article are excerpts from the example application. You can find the full source code on GitHub."
%}

### Login Page

{% include _image.liquid src="/assets/img/blogs/react-example-application/fusiondesk-login.png" alt="Login / Register Prompt if the user is not authenticated" class="img-fluid" figure=false %}

The login page allows the user to login or register using the FusionAuth login / register flow. The login page is implemented in the `LoginPage.tsx` file and displays the `login` and `register` buttons.

> We are not using the pre-built button components from the React SDK. Instead, we are using the `useFusionAuth` hook to get the `login` and `register` methods from the `FusionAuthContext`. This allows us to customize the login / register button.

Reviewer: `We are not using the pre-built button components from the React SDK.` - fine, but please explain why.

```tsx
  const {isAuthenticated, isLoading, login, register} = useFusionAuth();

  return (
    <div className="flex flex-col w-full border-opacity-50">
      <button onClick={() => login()} className="btn btn-primary">Login</button>
      <div className="divider">OR</div>
      <button onClick={() => register()} className="btn btn-primary">Register Now</button>
    </div>
  )
```

### Tickets Page

{% include _image.liquid src="/assets/img/blogs/react-example-application/fusiondesk-tickets.png" alt="Overview of the tickets when the user is authenticated" class="img-fluid" figure=false %}

> The tickets page displays the tickets of the user. The tickets are retrieved from the backend using the `/api/tickets` endpoint. The returned tickets depend on the role of the user. If the user is only a `customer`, only the tickets of the user are returned.

Reviewer: This would be good code to show.

We also embed the user information in the backend response. This allows us to display the creator information of the ticket (full name, picture).

```tsx
<div className="flex items-center space-x-3">
  <Avatar name={ticket._creator.firstName + ' ' + ticket._creator.lastName} url={ticket._creator.imageUrl}/>
  <div>{ticket._creator.firstName} {ticket._creator.lastName}</div>
</div>
```

> ### Ticket Page

Reviewer: ```suggestion
### Ticket details page
```

{% include _image.liquid src="/assets/img/blogs/react-example-application/fusiondesk-ticket.png" alt="Ticket details" class="img-fluid" figure=false %}

The ticket page allows the user to create or update a ticket. The ticket page is implemented in the `TicketPage.tsx` file.
Depending on the state of the ticket and the role of the user, different actions are available.

Currently, the following flow is implemented:

> {% include _image.liquid src="/assets/img/blogs/react-example-application/fusiondesk-flow.png" alt="The ticket flow inplemented in FusionDesk" class="img-fluid" figure=false %}

Reviewer: This image is missing.

If the ticket is open and the user is an `agent`, the user can mark the ticket as `solved`.

The creator of the ticket then has the option to either close or reopen the ticket with `Accept solution` or `Reject solution`.

After the ticket is closed, it cannot be edited anymore.

### Profile Page

{% include _image.liquid src="/assets/img/blogs/react-example-application/fusiondesk-profile.png" alt="Profile screen with information about the logged in user" class="img-fluid" figure=false %}

> The profile page displays the user information. The profile page is implemented in the `ProfilePage.tsx` file and displays the information of the FusionAuth `user`.

Reviewer: For each of these pages, it makes sense to show the back end API and then the front end JS.

      <figure><img src={user.picture} className="w-96 h-96" alt="Avatar"/></figure>
    ) : (
      <figure className="bg-neutral-focus text-neutral-content w-96 h-96">
        <span className="text-8xl">{initials(user.given_name + ' ' + user.family_name)}</span>
      </figure>
    )}
    <div className="card-body">
      <h2 className="card-title">{user.given_name} {user.family_name}</h2>
    
      <table>
        <tbody>
        <tr>
          <td><FontAwesomeIcon icon={faAt}/></td>
          <td>{user.email}</td>
        </tr>
        {user.phone_number &&
          <tr>
            <td><FontAwesomeIcon icon={faPhone}/></td>
            <td>{user.phone_number}</td>
          </tr>
        }
        </tbody>
      </table>
    </div>
  )
```

## Conclusion

In this article, we have shown how to use the FusionAuth React SDK in a React application. We have also shown how to use the FusionAuth OAuth 2.0 login flow in a React application and what the benefits are.

To get started with your own React application, install the FusionAuth React SDK and follow the [FusionAuth React SDK](https://fusionauth.io/docs/v1/tech/client-libraries/react-sdk) documentation.

```bash

> npm install @fusionauth/react-sdk

Reviewer: This should have been done at the start, rather than in the conclusion.

```bash
npm install @fusionauth/react-sdk
```

### Prerequisites

To follow along with this article, you will need to have the following installed:

* Node.js
* Git
* Docker with Docker Compose (optional)

{% include _callout-tip.liquid
content=
"If you want to run the example application on another domain than `localhost`, you need to make sure that FusionAuth runs on the same domain as the backend of the application. This is a limitation of the Hosted OAuth Service Provider Endpoints."
%}

{% include _callout-important.liquid
content=
"This article was written using an early look at the FusionAuth Hosted OAuth Service Provider Endpoints. The feature may be not yet available in the latest version of FusionAuth. You can track the progress at [Github fusionauth-issues#1943](https://github.com/FusionAuth/fusionauth-issues/issues/1943)"
%}

### Clone the Repository

To begin, clone the demo application from GitHub using the following command:

```bash
git clone https://github.com/sonderformat-llc/fusionauth-example-react-fusiondesk
```

### Set up the FusionAuth Application

To set up the FusionAuth application, follow the steps below:

```bash
cd fusionauth-example-react-fusiondesk

> docker compose up -d

Reviewer 2: Also, `docker-compose.yml` is pointing to a `fusionauth/fusionation-app:boostrap-dev` image. Shouldn't it be `fusionauth/fusionauth-app:latest`? It's kind of different of [the default compose file](https://github.com/FusionAuth/fusionauth-containers/tree/master/docker/fusionauth/docker-compose.yml)

[docker-compose.diff.txt](https://github.com/FusionAuth/fusionauth-site/files/11388686/docker-compose.diff.txt)

# Wait until FusionAuth is set up and available at http://localhost:9011
cd server && cp example.env .env && npm install && npm run seed && cd ..

> cd client && cp example.env .env && npm install && cd ..

Reviewer 2: I had some trouble with the 0.24.0 version of our React SDK, so I had to upgrade it:

```
diff --git a/client/package.json b/client/package.json
index d108474..bbd974b 100644
--- a/client/package.json
+++ b/client/package.json
@@ -6,7 +6,7 @@
     "@fortawesome/fontawesome-svg-core": "6.3.0",
     "@fortawesome/free-solid-svg-icons": "6.3.0",
     "@fortawesome/react-fontawesome": "0.2.0",
-    "@fusionauth/react-sdk": "0.24.0",
+    "@fusionauth/react-sdk": "^0.25.0",
     "@testing-library/jest-dom": "5.16.5",
     "@testing-library/react": "13.4.0",
     "@testing-library/user-event": "13.5.0",
```

Reviewer 2: And I had to change [server/src/middlewares/fusion-auth.middleware.ts](https://github.com/sonderformat-llc/fusionauth-example-react-fusiondesk/blob/main/server/src/middlewares/fusion-auth.middleware.ts#L31) to reflect the change in the cookie name.

```diff
diff --git a/server/src/middlewares/fusion-auth.middleware.ts b/server/src/middlewares/fusion-auth.middleware.ts
index 3509b14..a55f84d 100644
--- a/server/src/middlewares/fusion-auth.middleware.ts
+++ b/server/src/middlewares/fusion-auth.middleware.ts
@@ -28,7 +28,7 @@ export class FusionAuthMiddleware implements MiddlewareMethods, OnInit {
     }
 
     async use(@Context() $ctx: Context, @Locals() locals: any) {
-        const {access_token} = $ctx.request.cookies;
+        const access_token = $ctx.request.cookies['app.at'];
 
         if (access_token) {
             try {
```

```

### Run the Application

To run the application, follow the steps below:

```bash
cd server && npm run start
```

Open a new terminal and run the following command:

```bash
cd client && npm run start
```

> The application is now running on `http://localhost:3000`.

Reviewer 2: ```suggestion
The application is now running on [localhost:3000](http://localhost:3000).
```

To use the button styling provided by DaisyUI instead of the one provided by FusionAuth, we are not using the pre-built buttons from the React SDK. Instead, we are using the `useFusionAuth` hook to get the `login` and `register` methods from the `FusionAuthContext`.

```tsx
{% remote_include 'https://raw.githubusercontent.com/sonderformat-llc/fusionauth-example-react-fusiondesk/main/client/src/pages/LoginPage.tsx' %}
```

### Tickets Page

{% include _image.liquid src="/assets/img/blogs/react-example-application/fusiondesk-tickets.png" alt="Overview of the tickets when the user is authenticated" class="img-fluid" figure=false %}

The tickets page displays the tickets of the user. The tickets are retrieved from the backend using the `/api/tickets` endpoint. The returned tickets depend on the role of the user. If the user is only a `customer`, only the tickets of the user are returned.

We also embed the user information in the backend response. This allows us to display the creator information of the ticket (full name, picture).

Backend `GET /` endpoint:

> ```typescript

Reviewer 2: Maybe we could have endpoints in separated files so we can use `remote_include` here as well?
